src/python/pants/engine/fs.py

```diff
@@ -66,12 +66,26 @@ def __repr__(self) -> str:
         )
 
 
+@dataclass(frozen=True)
+class Directory:
+    """The path to a directory."""
+
+    path: str
+
+    def __repr__(self) -> str:
+        return f"Directory({self.path})"
```

Comment:
Is this to avoid quoting as done with `str`'s `__repr__`?

Reply:
It's to avoid `Directory(path="a/b/c")`, which is redundant. `Directory("a/b/c")` expresses it better. But actually, I did want the quotes, and this won't have it.

--

Also, I think that implementing `__repr__` will change `__str__`. Typically, `__str__` delegates to `__repr__`. Not sure if that stays true with dataclasses.

Reply:
It does!! Using my favorite feature of I think python 3.9 (f-strings with `{<expr>=}`):
```python
Python 3.10.0a0 (heads/python-object-scanner-dirty:268b849071, Jun 21 2020, 20:29:49)
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from dataclasses import dataclass
>>> @dataclass
... class A:
...   x: int
...   y: str
...
>>> repr(A(1, "a"))
"A(x=1, y='a')"
>>> str(A(1, "a"))
"A(x=1, y='a')"
>>> @dataclass
... class A:
...   x: int
...   y: str
...   def __repr__(self):
...     return f'new repr!!! {self.x=}, {self.y=}'
...
>>> repr(A(1, "a"))
"new repr!!! self.x=1, self.y='a'"
>>> str(A(1, "a"))
"new repr!!! self.x=1, self.y='a'"
```

Reply:
While I understand the desire to avoid rolling too much custom code to avoid python objects, if it allows us to do extremely simple things robustly and consistently (e.g. formatting without providing the name of the field), I think there *may* be an argument for a `@print_without_field_names` decorator?

Reply:
I don't think right now. If we see ourselves doing this enough is all. It's less space as writing a comment to explain it.